pipelined/bevy_render2/src/view/visibility/mod.rs

```diff
@@ -0,0 +1,187 @@
+        use VisibilitySystems::*;
+
+        app.add_system_to_stage(
+            CoreStage::PostUpdate,
+            calculate_bounds.label(CalculateBounds),
+        )
+        .add_system_to_stage(
+            CoreStage::PostUpdate,
+            update_frusta::<OrthographicProjection>
+                .label(UpdateOrthographicFrusta)
+                .after(TransformSystem::TransformPropagate),
+        )
+        .add_system_to_stage(
+            CoreStage::PostUpdate,
+            update_frusta::<PerspectiveProjection>
+                .label(UpdatePerspectiveFrusta)
+                .after(TransformSystem::TransformPropagate),
+        )
+        .add_system_to_stage(
+            CoreStage::PostUpdate,
+            check_visibility
+                .label(CheckVisibility)
+                .after(CalculateBounds)
+                .after(UpdateOrthographicFrusta)
+                .after(UpdatePerspectiveFrusta)
+                .after(TransformSystem::TransformPropagate),
+        );
+    }
+}
+
+pub fn calculate_bounds(
+    mut commands: Commands,
+    meshes: Res<Assets<Mesh>>,
+    without_aabb: Query<(Entity, &Handle<Mesh>), Without<Aabb>>,
+) {
+    for (entity, mesh_handle) in without_aabb.iter() {
+        if let Some(mesh) = meshes.get(mesh_handle) {
+            if let Some(aabb) = mesh.compute_aabb() {
+                commands.entity(entity).insert(aabb);
+            }
+        }
+    }
+}
+
+pub fn update_frusta<T: CameraProjection + Send + Sync + 'static>(
+    mut views: Query<(&GlobalTransform, &T, &mut Frustum)>,
+) {
+    for (transform, projection, mut frustum) in views.iter_mut() {
+        let view_projection =
+            projection.get_projection_matrix() * transform.compute_matrix().inverse();
+        *frustum = Frustum::from_view_projection(
+            &view_projection,
+            &transform.translation,
+            &transform.back(),
+            projection.far(),
+        );
+    }
+}
+
+pub fn check_visibility(
+    mut view_query: Query<(&mut VisibleEntities, &Frustum, Option<&RenderLayers>), With<Camera>>,
+    mut visible_entity_query: Query<(
+        Entity,
+        &Visibility,
+        &mut ComputedVisibility,
+        Option<&RenderLayers>,
+        Option<&Aabb>,
+        Option<&GlobalTransform>,
+    )>,
+) {
+    let mut first_view = true;
+    for (mut visible_entities, frustum, maybe_view_mask) in view_query.iter_mut() {
+        visible_entities.entities.clear();
+        let view_mask = maybe_view_mask.copied().unwrap_or_default();
+
+        for (
+            entity,
+            visibility,
+            mut computed_visibility,
+            maybe_entity_mask,
+            maybe_aabb,
+            maybe_transform,
+        ) in visible_entity_query.iter_mut()
+        {
+            // Initialize the computed visibility to false for the first view, then it is inherited for subsequent views
+            if first_view {
+                computed_visibility.is_visible = false;
```

Comment:
I think there should just be a simple ComputedVisibility query iterator that resets everything to false at the start of this system. That makes control flow simpler, and it also ensures that we reset computed visibility properly in the event that there are no views. Doing this will require using a QuerySet, but imo its worth it.

Reply:
Good point. I had it that way but felt it was probably important not to do two passes over the entities so tried to optimise it away. Good catch that if there were no views then it wouldn’t be reset.

Reply:
Done. (Mainly noting for myself. :))